virtual/registry/sql_kv.go

```diff
@@ -0,0 +1,194 @@
+package registry
+
+import (
+	"context"
+	"database/sql"
+	"errors"
+	"fmt"
+	"time"
+)
+
+// TODO: hardcoded for now, but may be useful as a configuration option for
+// some type of isolation
+const defaultTableName = "nola_kv"
+
+type sqlKV struct {
+	db *sql.DB
+	sqlKVStatement
+}
+
+type sqlKVStatement struct {
+	tableName  string
+	vst        time.Time
+	setStmt    *sql.Stmt
+	getStmt    *sql.Stmt
+	deleteStmt *sql.Stmt
+	rangeStmt  *sql.Stmt
+}
+
+func newSQLKV(driver, dsn string) (kv, error) {
+	tableName := defaultTableName
+	db, err := sql.Open(driver, dsn) // TODO support selecting driver at runtime
+	if err != nil {
+		return nil, err
+	}
+
+	db.SetMaxOpenConns(1)
+	db.SetMaxIdleConns(1)
+	db.SetConnMaxIdleTime(0)
+
+	err = db.Ping()
+	if err != nil {
+		return nil, err
+	}
+
+	// TODO Improve indexing
+	_, err = db.Exec("CREATE TABLE IF NOT EXISTS nola_kv (k BLOB PRIMARY KEY, v BLOB NOT NULL);")
+	if err != nil {
+		return nil, fmt.Errorf("failed to create table: %w", err)
+	}
+
+	setStmt, err := db.Prepare("INSERT INTO nola_kv (k, v) VALUES (?,?) ON CONFLICT (k) DO UPDATE SET v=?;")
+	if err != nil {
+		return nil, fmt.Errorf("failed to prepare insert statement: %w", err)
+	}
+
+	getStmt, err := db.Prepare("SELECT v FROM nola_kv WHERE k=?;")
+	if err != nil {
+		return nil, fmt.Errorf("failed to prepare select statement: %w", err)
+	}
+
+	deleteStmt, err := db.Prepare("DELETE FROM nola_kv where k=?;")
+	if err != nil {
+		return nil, fmt.Errorf("failed to prepare delete statement: %w", err)
+	}
+
+	rangeStmt, err := db.Prepare("SELECT k, v FROM nola_kv WHERE k LIKE ?;")
```

Comment:
Can you use any of these functions (sub string maybe?) to make this faster: https://www.postgresql.org/docs/9.0/functions-binarystring.html